### PR TITLE
[SPARK-32927][SQL] Bitwise OR, AND and XOR should have similar canonicalization rules to boolean OR and AND

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Add canonicalization rules for commutative bitwise operations.

### Why are the changes needed?

Canonical form is used in many other optimization rules. Reduces the number of cases, where plans with identical results are considered to be distinct. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala:84]: should we check whether both branches are deterministic like or/and in the above?

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala:84]: Bitwise operations will always execute both sides. Boolean AND and OR can short circuit and not execute one side.
The bitwise operations are more like + and * in this regard.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala:84]: OK. I was thinking more about side-effect from non-deterministic expressions (e.g., println), and the evaluation order matters in this case, but it seems other existing transformations don't handle this as well.

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
add to whiltelist

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala:101]: nit:

```scala
    Seq(BitwiseOr(_, _), BitwiseAnd(_, _), BitwiseXor(_, _)).foreach { f =>
      val e1 = f('a, f('b, 'c))
      val e2 = f(f('a, 'b), 'c)
      val e3 = f('a, f('b, 'a))

      assert(e1.canonicalized == e2.canonicalized)
      assert(e1.canonicalized != e3.canonicalized)
    }
```

Comment (tanelk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala:101]: Oh, didn't know that syntax, thank you, will change.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala:100]: Could you add more test cases, e.g., non-deterministic exprs, literals, and mixed cases `BitwiseOr`/`BitwiseAnd`/`BitwiseXor`?

**Review by maropu — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]
Looks fine to me too

**Comment by maropu** [comment]
retest this please

**Comment by tanelk** [comment]
@maropu , @HyukjinKwon are there any more issues to solve before this can be merged?

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @viirya from https://github.com/apache/spark/pull/15388. Will merge this in if there are no more comments from you.

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala:84]: Canonicalized expression is only used for determining if two expressions have same result. We don't run it directly. And we usually consider `deterministic` together when using `canonicalized`, e.g. `semanticEquals`.

**Comment by maropu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
